File: [백준]진우의민트초코우유.py
# 체력은 진우가 이동할 수 있는 거리
# 민트초코우유를 마신다면 체력이 H 만큼 증가
# 최대한 우유 많이먹고 집에 돌어오는(중간에 체력0되지않을만큼)
import copy
import sys
sys.stdin = open(".\\input.txt")
input = sys.stdin.readline
#마을크기,초기체력,증가체력
N,M,H = list(map(int,input().split()))
#1=집,2=우유
board = []
MILK = []
for i in range(N):
    li = list(map(int,input().split()))
    board.append(li)
    if 1 in li:#집의 위치
        start = tuple([i,li.index(1)])
    for j in range(N):
        if li[j] == 2:
            MILK.append([i,j])
ans = 0
nb = copy.deepcopy(board)
milk = copy.deepcopy(MILK)
def dfs(milk,nb,now,M):#먹은 우유개수, 먹은게 갱신된 지도, 현위치, 남은체력
    #체력이 닿는 곳 까지 다 돌았을 때 milk있는것들 다 담아야지
    global ans
    for m in milk:
        move = abs(m[0]-now[0]) +abs(m[1]-now[1])
        if  move<= M:
            M = M-move+H#현재체력-이동량+우유먹은체력
            if M>=0:#체력이 남아있으면
                now = m#현위치 바꾸고
                milk.remove(m)#먹었으니까 제거
                nb[m[0]][m[1]] = 3
                dfs(milk,nb,now,M)
                nb[m[0]][m[1]] = board[m[0]][m[1]]
        if now!=start and abs(start[0]-now[0]) +abs(start[1]-now[1]) <=M:#현위치가 집이 아니고 집까지돌아갈수있으면
            ans = max(ans,len(MILK)-len(milk))#원래우유개수-못먹은우유개수
            now = start#집으로 돌아옴
            

dfs(milk,nb,start,M)
print(ans)
#output:2,8

# 일단 갈 수 있는 모든 경우의수를 q에 넣고 해당 지점에서 집으로 돌아올수있으면 돌아오고
# 집으로 가는 여부와 상관없이 다른 지점으로 갈수 있으면 간다. 

File: [백준]진우의민트초코우유.cpp
//파이썬으로는 시간초과를 피할수 없는문제
//(진짜 최적화를 잘 시킨다면 가능이야하겠지만 그만큼의 알고리즘을 요구하는게 아닌 문제)
#include <iostream>
#include <vector>
#include <algorithm>
#include <fstream>
using namespace std;
using pr = pair<int, int>;

int board[11][11];
int ans = 0;
int N, M, H;
int milk_cnt;
pr start;

void dfs(vector<pr> milk, pr now, int M) {
   vector<pr> mk;
   pr m;
   
   int move,nM;
   for (int i = 0; i < milk.size(); i++) {
      mk.assign(milk.begin(), milk.end());//milk를 mk에 깊은 복사

      m = milk[i];
      move = abs(m.first - now.first) + abs(m.second - now.second);
      if (move <= M) {
         nM = M - move + H;
         if (nM >= 0) {
            auto it = remove(mk.begin(), mk.end(), m);
            mk.resize(it - mk.begin());
            dfs(mk, m, nM);
            if (m != start) {
               if (abs(start.first - m.first) + abs(start.second - m.second) <= nM) {//왜냐면 현재 now는 m임
                  ans = ans >= milk_cnt - mk.size() ? ans : milk_cnt - mk.size();
               }
            }
         }
      }
   }
}
int main() {
   ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
   //freopen(".\\input.txt", "r", stdin);
   cin >> N >> M >> H;
   vector<pr> MILK;
   for (int i = 0; i < N; i++) {
      for (int j = 0; j < N; j++) {
         cin >> board[i][j];
         if (board[i][j] == 1) {
            start.first = i;
            start.second = j;
         }
         else if (board[i][j] == 2) {
            MILK.push_back({ i,j });
         }
      }
   }
   milk_cnt = MILK.size();
   dfs(MILK, start, M);
   cout << ans;
}